### ISSUE TITLE
Parser Bug: For Loop Fails When 'in' Keyword Is Followed by a Line Break and an Array

### ISSUE THREAD

**Issue by Enter-tainer** [typst/typst#4548]
### Description

The parser fails to correctly handle a for loop when the 'in' keyword is followed by a line break and the iterable is an array. This results in compilation errors.

```typst
#{
for i in 
  (1, 2, 3) {
  [#i]
}
}

```

```
Error: expected expression
   ╭─[/main.typ:9:9]
   │
 9 │ for i in
───╯

Error: expected semicolon or line break
    ╭─[/main.typ:10:12]
    │
 10 │   (1, 2, 3) {
────╯

```

### Operating system

Web app, Windows, Linux, macOS

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enter-tainer** [comment]
Also for dict:

```typst
#{
for i in
  (a: 1, b: 2) {
  [#i]
}
}
```

**Comment by Leedehai** [comment]
I'm not sure if this is a "bug", or a WAI? Maybe @laurmaedje

**Comment by wrzian** [comment]
In a code block this would not be expected. I believe this is a bug, looking into it.

**Comment by Enter-tainer** [comment]
agree. For markup i think it is intended. But for code i think it's a bug

**Comment by Enter-tainer** [comment]
any idea why this is not allowed in code mode? This enforce binary op chains should always stay in the same line.

```
#{
let a = 1
 + 2
 + 3
  [a]
}
```

```
Error: cannot join integer with integer
    ╭─[/main.typ:11:2]
    │
 11 │  + 3
────╯
```

```
#{
let a = 1 + 
  2 +
  3
  [a]
}

```
```
Error: expected expression
   ╭─[/main.typ:9:12]
   │
 9 │ let a = 1 +
───╯

Error: expected expression
    ╭─[/main.typ:10:6]
    │
 10 │   2 +
────╯

```

**Comment by laurmaedje** [comment]
> In a code block this would not be expected. I believe this is a bug, looking into it.

If no parentheses are placed, even in code mode there are only two cases (IIRC) where a line can continue: With a full stop for field access or method call, or with an `else`. Everything else cannot continue. We could potentially expand this to `in` in a for loop. I wouldn't classify it as a bug, but rather as a design choice that we can discuss.

**Comment by wrzian** [comment]
It's definitely not a bug, but I'll argue that it should be changed because it is unintuitive and limits code flexibility where we don't need to, effectively enforcing a style guide on code that is more of a job for a linter.

## Current Behavior

First let's clarify the current behavior.

### Non code-blocks
```typ
Normal code eagerly converts to markup and only
allows "atomic" expressions:
#(5+2)+4
// a. => "7+4"

Although some common statements are "non-atomic" and
allow binary operators etc:
#let x = (5 + 2) + 4
#x
// b. => "9"

Expressions like `(x) => x + 1` or `context` inherit
their atomicity from their parent:
#let y = context x + 1
#y
// c. => "10"
#context x + 1
// d. => "9+ 1"
```

### Code-blocks

In code-blocks, two things change:
1. We allow lines to span expressions only in the case of dot-continuations and else-statements. This is hard-coded in the current `Parser::lex` function, although I've recently moved it into the Lexer in #4364 as part of the refactor.
2. We allow expressions on the same line, but only if they're separated with a semicolon.

Otherwise we keep the same newline behavior, which causes the above for-loop behavior.

```typ
#if false [ True ]
else [ False ]
// e. => "else [ False ]"
#{
  if false [ True ]
  else [ False ]
} // f. => "False"

#{
  (1, 2, 3)
    .map(x => x + 1)
    .filter(x => x > 2)
} // g. => (3, 4)

#{
  for i in
  (1, 2, 3) { i }
} // h. => Error! No expression after 'in'
#{
  let y = (x)
  => x + 1
} // i. => Error! Unexpected arrow '=>'


#let naive_result(result /*dict like: (ok: {}, err: {})*/) = {
  result.ok
  if result.err != none {
    return
    result.err
  }
}
#naive_result((ok: "ok", err: none))
// j. => "ok"
#naive_result((ok: "still ok", err: "some error"))
// k. => "still ok"
// Wait what?!
This one is tricky. It's syntactically allowed, but
a return followed by a newline doesn't include the
newline expression and returns any existing placed
content.
```

This also has a really annoying interaction with number literal syntax. Try to guess if this produces content or an error, and for either, what?
```typ
#{
  let x = 5
  + 4
  #x
} // l.
```
<details><summary>Answer</summary>
<p>

The answer is `"45"`!

This comes the use of leading minus for negation (`-4`) and the symmetry that allows leading plus in `+4` combined with whitespace insensitivity. Lovely.

</p>
</details> 

## Ok, Now What?

First, some prior art.

```py
# Python disallows this:
x = 1 +
2
# py. => syntax error
```
```js
// But it works in Javascript:
var x = 1 +
2
// js. => x === 3
```

### Options

I see three likely options for code-blocks in Typst.

1. **Greedy newlines:** Keep the current behavior of greedily checking for newlines in the lexer, but carve-out specific edge cases for common items like if-else or trailing-dot function calls.

2. **Greedy expressions:** Expressions try to operate over whitespace if possible within blocks and parentheses, stopping only at explicit semicolons. This would change examples `g` through `l`. And would affect leading minus signs at the start of lines (usually meaning negation to now usually meaning subtraction with previous token), but I think this could work well.

3. **Indentation sensitivity:** An alternative is a hybrid approach, where we require binary operators continuing over lines to to be indented further than the starting column of their parent in the source code. This is an odd choice, but does have prior art in Haskell and some Lisp derived languages (see i.e. some [Rhombus lang syntax exporations](https://github.com/mflatt/rhombus-prototype/blob/shrubbery/shrubbery/0000-shrubbery.md#continuing-with-indentation-and-an-operator))

<details><summary><b>Indentation sensitivity</b> examples</summary>
<p>

We might allow this trailing arrow so long as it increases the indenation of the previous statment sufficiently. There are a few bikeshedding opportunities here, but the general idea should stay the same.
```typ
#{
  let x = (really, long, chain, of, arguments)
    => { ... }
  // m. => No error
  
  let y = 1
  + 2
  #y
  // n.  => "21" (see above)

  let z = 1
    + 2
  #z
  // o. => "3"
}
```

</p>
</details> 

### Pros/Cons

**Greedy newlines** is the best for error messages and performance: it makes it easy to stop parsing at newlines and generate partial parses in an IDE for separating errors. One notable win is if you type a bare `let` with no variable in the middle of a code-block, the parser only has to generate an error on that line and doesn't have to search forward for a potential identifier and equals-sign to associate.

It is also probably good for beginners as it imposes the most requirements on syntax so users can't type things incorrectly and error messages are usually localized if they do happen.

**Greedy expressions** is the most ergonomic and flexible. It's good for code formatting as expressions can be moved arbitrarily and makes the parser's implementation much cleaner. I'm personally a fan of this one, but it does increase the scope of possible errors and likely some performance burden. 

**Indentation sensistivity** is the most consistent and likely most beautiful. With **greedy expressions**' comes flexibility for beauty but also flexibility for ugliness. Indentation sensitivity makes using binary operators over multiple lines require explicit indentation which clarifies programmer intent. However this is likely the worst for teaching newcomers and would complicate the existing parser.

|  | Errors | Teaching | Flexibility | Beauty | 
|--------|--------|--------|--------|--------|
| **Greedy newlines** | ✅ | ✅ | ❌ | ❔ |
| **Greedy expressions** | ❌ | ✅ | ✅ | ❔ |
| **Indentation sensistivity** | ❔ | ❌ | ❔ | ✅ |

## Conclusion

I'm personally partial to **greedy expressions**, it's the most flexible syntactically and feels the most like a "real programming language". We can also make choices at a more fine-grained level (like requiring return expressions to be on the same line), but I think it's important to have a well-defined default method first.

We should also expand on these nuances in the syntax documentation and the Parser. I'll make a new issue for that, might do it myself, but would appreciate help if anyone's interested in writing things up.

And again, this discussion only applies to code syntax *within code-blocks*. Regardless of what is chosen for code-blocks, I think we should absolutely keep the **greedy newline** behavior with the current atomic/non-atomic system when not in code-blocks.
